### PR TITLE
Add programming assessment dashboard

### PR DESCRIPTION
**Pull request by SimonEntholzer**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Unlike modeling and textexercises, there was no assessment-dashboard for programming exercises. In [#2745](https://github.com/ls1intum/Artemis/pull/2745) we adapted the dashboards for the second correction. The dashboard we create here should work in the same way as those.

### Description
You can access the dashboard from the exercise groups page, when you click on "Submission".
image
There you will see a similar dashboard as with other exercise types.
image

### Steps for Testing

1. create an exam with 2 correction rounds, and with a programming exercise
2. participate with a few students (2-5)

Now, the easiest way to test, is with two users simultaneously in two different browser windows. With one tutor account go to the exercise-assessment dashboard. With the instructor account go to the new assessment dashboard. (exam -> exercise-groups -> programming-exercise -> submissions)

First the assessment dashboard should show submissions only with automatic results. Now start an assessment with the tutor from the exercise-assessment-dashboard.
 

Refresh the assessment-dashboard, now one locked submission should be visible. Submit with the tutor, now it should be visible with "open assessment" from the instructors perspective. When you open it with the instructor, make sure everything is loaded correctly.

With a different tutor do the second correction, and test the same things again. (You might need to enable the second correction with the instructor in the exercise-assessment-dashboard).

Now use the assessment dashboard with the instructor to assess different submissions in the first/second round.

When opening results with "open/continue assessment" be sure that you get the result that you would expect.

### Test Coverage

programming-assessment-dashboard.component.ts: 0% (Not existing yet, I will create it before the merge)

### CODE REVIEW

**Review by airelawaleria — DISMISSED** [review]
worked as expected on ts1

**Review by JohannesWeiss — APPROVED** [review]
Tested on ts1 and works as intended. Nice feature! 👍

**Review by simonzilker — DISMISSED** [review]
Tested with @marvinchimeka . Worked as described

**Review by marvinchimeka — DISMISSED** [review]
tested in testing session. Could not find any issues.

**Review by EneaGore — DISMISSED** [review]
LG on ts1

**Review by philgabr — APPROVED** [review]
Works as described.

**Review by LudwigWeide — DISMISSED** [review]

**Review by tiffca — APPROVED** [review]
Works as expected, no issues found.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 1]
```diff
@@ -141,7 +141,7 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
         }
 
         if (!programmingExercise.areManualResultsAllowed()) {
-            return forbidden();
+            return forbidden("assessment", "assessmentSaveNotAllowed", "Creating nanual results is disabled for this exercise!");
```
Comment (nicolasruscher) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:144]: I think this should be 'manual' :D

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java [hunk 2]
```diff
@@ -110,6 +110,23 @@ public ProgrammingExerciseParticipationResource(ProgrammingExerciseParticipation
         // Set exercise back to participation
         participation.get().setExercise(exercise);
 
+        // get the result which belongs to the specific correctionround and save it as the single result in the participation
```
Comment (nicolasruscher) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java:113]: I think this comment is misleading, as it is not really being saved is it? Maybe 'set' would be better here

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java [hunk 3]
```diff
@@ -389,8 +390,9 @@ public ProgrammingSubmissionResource(StudentParticipationRepository studentParti
             return forbidden();
         }
 
-        Optional<Result> manualResult = participation.getResults().stream().filter(Result::isManual).findFirst();
-        if (manualResult.isPresent()) {
+        // Optional<Result> manualResult =
```
Comment (nicolasruscher) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:393]: This comment can be removed

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts [hunk 4]
```diff
@@ -148,9 +148,10 @@ export class CodeEditorTutorAssessmentContainerComponent implements OnInit, OnDe
             } else {
                 participationId = Number(params['participationId']);
             }
+            // correctionRound = Number(params['correctionRound']);
```
Comment (nicolasruscher) [inline src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts:151]: Can this be removed ? :D

**Review by nicolasruscher — DISMISSED** [review]
Looked over the code, mainly the server side changes, looks good imo apart from a few comments that should be removed, good job 👍 Currently the client jobs are missing tho, will they be added in a followup?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java (hunk 1 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:144]: oh yea :D 
```suggestion
            return forbidden("assessment", "assessmentSaveNotAllowed", "Creating manual results is disabled for this exercise!");
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java (hunk 2 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java:113]: you are right, we don't save it to the database here
```suggestion
        // get the result which belongs to the specific correctionround and set it as the single result in the participation
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java (hunk 3 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:393]: yes, I'll remove it
```suggestion

```

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts (hunk 4 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts:151]: I removed this commented code too :)
```suggestion

```

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html [hunk 5]
```diff
@@ -0,0 +1,181 @@
+<div *ngIf="exercise">
+    <h4 [translateValues]="{ exerciseTitle: exercise.title }" jhiTranslate="artemisApp.programmingAssessment.dashboard.heading"></h4>
+    <jhi-assessment-filters (filterChange)="updateFilteredSubmissions($event)" [submissions]="submissions"></jhi-assessment-filters>
+    <ng-container *ngIf="!busy"></ng-container>
```
Comment (iozbeyli) [inline src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html:4]: There is nothing inside this ng-container, am I missing something?

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+                    </td>
+                    <td>{{ assessmentTypeTranslationKey(submission.latestResult) | translate }}</td>
+                    <td>
+                        <span *ngIf="submission.latestResult && submission.latestResult!.assessor">{{ submission.latestResult!.assessor!.name }}</span>
+                    </td>
+                    <td *ngFor="let item of [].constructor(numberOfCorrectionrounds); let correctionRound = index">
+                        <span *jhiHasAnyAuthority="['ROLE_ADMIN', 'ROLE_INSTRUCTOR', 'ROLE_TA']">
+                            <!-- exercise without exam-->
+                            <span *ngIf="exercise.type === ExerciseType.PROGRAMMING && exercise.course && submission">
+                                <button *ngIf="submission.results.length == 0" [disabled]="busy" [routerLink]="getAssessmentLink(submission)" class="btn btn-success btn-sm mb-1">
```
Comment (iozbeyli) [inline src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html:74]: There was a pull request that changes buttons onto links, so wouldn't it be nice to follow that approach in this pull request as well.

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.module.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        FormDateTimePickerModule,
+        FormsModule,
+        FeatureToggleModule,
+        AssessmentInstructionsModule,
+        RouterModule.forChild([]),
+        ArtemisResultModule,
+        ArtemisAssessmentSharedModule,
+    ],
+    declarations: [ProgrammingAssessmentDashboardComponent],
+    entryComponents: [],
```
Comment (iozbeyli) [inline src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.module.ts:24]: These empty arrays are not necessary.

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html (hunk 5 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html:4]: This part I just copied from the other _exercisetype_-assessment-dashboards. As it turns out in a previous version of this code (in the text-assessment-dashboard e.g.) there was an else branch if the condition was true. I will delete it.

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.module.ts (hunk 7 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.module.ts:24]: I removed those too, thanks for the tip :D

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html (hunk 6 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html:74]: I changed all the buttons to links :)

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+                                <a
+                                    *ngIf="submission.results.length == 1 && !submission.latestResult.completionDate"
+                                    [class.disabled]="busy"
+                                    [routerLink]="getAssessmentLink(submission)"
+                                    class="btn btn-warning btn-sm mb-1"
+                                >
+                                    <fa-icon [fixedWidth]="true" [icon]="'folder-open'"></fa-icon>&nbsp;{{ 'artemisApp.assessment.dashboard.actions.continue' | translate }}
+                                </a>
+                                <a
+                                    (click)="cancelAssessment(submission)"
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html:93]: This doesn't use a router link but fires an `click` event, so it should remain a button.

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                                        >
+                                            <fa-icon [fixedWidth]="true" [icon]="'edit'"></fa-icon>&nbsp;
+                                            <span>
+                                                {{ 'artemisApp.assessment.dashboard.actions.continue' | translate }}
+                                            </span>
+                                        </a>
+                                        <!-- button to cancel result -->
+                                        <span *ngIf="exercise.isAtLeastInstructor">
+                                            <a
+                                                (click)="cancelAssessment(submission)"
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html:139]: Same here.

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html (hunk 8 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html:93]: ah yea you are right :) I changed all buttons and did indeed overlook this case. I'll change it back.

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
I would remove the "Score" column and just use the latest result column for sorting (similar to the Scores view). This would save space as the information can be found in the latest result column as well.

Is there any reason why we would like to have an additional "Score" column that I didn't think of?

**Review by fde312 — CHANGES_REQUESTED** [review]
When a exercise has not a the manual assessment enabled, you still can enter the assessment view and lock the submission.
image

image

Please add here a `ngIf` to not display the button when the exercise has no manual assessment enabled.
We should definitely add also a check in the REST Call for locking the submission to catch this issue as well (you can do it here or I add it in a follow up).

Other than that everything worked fine for me! :)

**Review by MaisiKoleni — COMMENTED** [review]
Works well apart from the issue Francisco mentioned. The exam correction rounds feature seems to work well 👍

**Comment by SimonEntholzer** [comment]
> When a exercise has not a the manual assessment enabled, you still can enter the assessment view and lock the submission.
> image
> 
> image
> 
> Please add here a `ngIf` to not display the button when the exercise has no manual assessment enabled.
> We should definitely add also a check in the REST Call for locking the submission to catch this issue as well (you can do it here or I add it in a follow up).
> 
> Other than that everything worked fine for me! :)

I added checks to not show the buttons if manual assessment is disabled for the exercise.

File: src/main/webapp/app/exercises/modeling/participate/modeling-submission.service.ts [hunk 10]
```diff
@@ -90,6 +90,7 @@ export class ModelingSubmissionService {
     /**
      * Get a submission with given Id
      * @param {number} submissionId - Id of the submission
+     * @param correctionRound
```
Comment (derLalla) [inline src/main/webapp/app/exercises/modeling/participate/modeling-submission.service.ts:93]:
```suggestion
     * @param {correctionRound}
```

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+                                            [routerLink]="getAssessmentLink(submission)"
+                                            class="btn btn-primary btn-sm mb-1"
+                                        >
+                                            <fa-icon [fixedWidth]="true" [icon]="'folder-open'"></fa-icon>&nbsp;
+                                            <span>
+                                                {{ 'artemisApp.assessment.dashboard.actions.open' | translate }}
+                                            </span>
+                                        </a>
+                                        <!-- button for existing result with lock-->
+                                        <span *ngIf="submission.results && !submission?.results[correctionRound].completionDate">
```
Comment (derLalla) [inline src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html:123]:
```suggestion
                                        <span *ngIf="submission.results && !submission.results[correctionRound]?.completionDate">
```

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+                                            </a>
+                                            <!-- button to cancel result -->
+                                            <span *ngIf="exercise.isAtLeastInstructor">
+                                                <button
+                                                    (click)="cancelAssessment(submission)"
+                                                    *ngIf="
+                                                        submission &&
+                                                        submission.latestResult &&
+                                                        submission.latestResult.assessmentType !== automaticType &&
+                                                        !submission.latestResult!.completionDate
```
Comment (derLalla) [inline src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.html:143]:
```suggestion
                                                        !submission.latestResult.completionDate
```

File: src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.ts [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+            .getProgrammingSubmissionsForExerciseByCorrectionRound(this.exercise.id!, { submittedOnly: true })
+            .map((response: HttpResponse<ProgrammingSubmission[]>) =>
+                response.body!.map((submission: ProgrammingSubmission) => {
+                    const tmpResult = getLatestSubmissionResult(submission);
+                    setLatestSubmissionResult(submission, tmpResult);
+                    if (tmpResult) {
+                        // reconnect some associations
+                        tmpResult!.submission = submission;
+                        tmpResult!.participation = submission.participation;
+                        submission.participation!.results = [tmpResult!];
```
Comment (derLalla) [inline src/main/webapp/app/exercises/programming/assess/programming-assessment-dashboard/programming-assessment-dashboard.component.ts:85]:
```suggestion
                    if (tmpResult) {
                        // reconnect some associations
                        tmpResult.submission = submission;
                        tmpResult.participation = submission.participation;
                        submission.participation!.results = [tmpResult];
```

**Review by derLalla — CHANGES_REQUESTED** [review]
code looks good. I have just noted some minor changes :)

**Review by ivanchimeno — APPROVED** [review]
Approving because testing worked on TS3 and @derLalla found the issues that need to be addressed.

**Review by derLalla — APPROVED** [review]
re-approving code changes

File: src/main/webapp/app/exercises/shared/exercise/exercise-utils.ts [hunk 14]
```diff
@@ -159,7 +159,11 @@ export const areManualResultsAllowed = (exercise: Exercise) => {
     // Only allow new results if manual reviews are activated and the due date/after due date has passed
     const progEx = exercise as ProgrammingExercise;
     const relevantDueDate = progEx.buildAndTestStudentSubmissionsAfterDueDate ?? progEx.dueDate;
-    return progEx.isAtLeastTutor === true && progEx.assessmentType === AssessmentType.SEMI_AUTOMATIC && (!relevantDueDate || moment(relevantDueDate).isBefore(now()));
+    return (
+        (progEx.isAtLeastTutor === true || progEx.isAtLeastInstructor === true) &&
```
Comment (SimonEntholzer) [inline src/main/webapp/app/exercises/shared/exercise/exercise-utils.ts:163]:
```suggestion
        (progEx.isAtLeastTutor || progEx.isAtLeastInstructor) &&
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java [hunk 15]
```diff
@@ -389,8 +390,12 @@ public ProgrammingSubmissionResource(StudentParticipationRepository studentParti
             return forbidden();
         }
 
-        Optional<Result> manualResult = participation.getResults().stream().filter(Result::isManual).findFirst();
-        if (manualResult.isPresent()) {
+        if (!((ProgrammingExercise) participation.getExercise()).areManualResultsAllowed()) {
```
Comment (fde312) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:393]: Thank you :)

**Review by ivanchimeno — APPROVED** [review]
Re-approve

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java (hunk 15 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:393]: you're welcome :D

**Review by fde312 — APPROVED** [review]
Now it works for me:
image
thanks also for the server check 👍 

Really cool feature! 🚀

**Review by DenizShabani — DISMISSED** [review]

File: src/main/webapp/app/exercises/shared/exercise/exercise-utils.ts [hunk 16]
```diff
@@ -159,7 +159,11 @@ export const areManualResultsAllowed = (exercise: Exercise) => {
     // Only allow new results if manual reviews are activated and the due date/after due date has passed
     const progEx = exercise as ProgrammingExercise;
     const relevantDueDate = progEx.buildAndTestStudentSubmissionsAfterDueDate ?? progEx.dueDate;
-    return progEx.isAtLeastTutor === true && progEx.assessmentType === AssessmentType.SEMI_AUTOMATIC && (!relevantDueDate || moment(relevantDueDate).isBefore(now()));
+    return (
+        (progEx.isAtLeastTutor || progEx.isAtLeastInstructor) &&
```
Comment (SimonEntholzer) [inline src/main/webapp/app/exercises/shared/exercise/exercise-utils.ts:163]:
```suggestion
        (!!progEx.isAtLeastTutor || !!progEx.isAtLeastInstructor) &&
```

**Review by ivanchimeno — DISMISSED** [review]
Re-approve

**Review by fde312 — DISMISSED** [review]
reapprove

File: src/main/webapp/app/exercises/shared/exam-exercise-row-buttons/exam-exercise-row-buttons.component.html [hunk 17]
```diff
@@ -96,6 +96,7 @@
 
     <a
         *ngIf="course.isAtLeastInstructor && exercise.type !== exerciseType.QUIZ"
+        type="submit"
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/shared/exam-exercise-row-buttons/exam-exercise-row-buttons.component.html:99]: This doesn't look correct, see https://github.com/ls1intum/Artemis/pull/2669#discussion_r568288094.

**Review by FrankeLukas — DISMISSED** [review]
Lgtm otherwise

File: src/main/webapp/app/exercises/shared/exam-exercise-row-buttons/exam-exercise-row-buttons.component.html (hunk 17 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/shared/exam-exercise-row-buttons/exam-exercise-row-buttons.component.html:99]:
```suggestion
```

**Review by FrankeLukas — APPROVED** [review]
Re-approve.

**Review by fde312 — APPROVED** [review]

**Review by ivanchimeno — APPROVED** [review]
:rocket:

**Review by derLalla — APPROVED** [review]
reapproving code changes 👍

**Review by krusche — APPROVED** [review]

**Review by FuchsDominik — APPROVED** [review]
Looks good to me :)

**Pull request merged**